### PR TITLE
feat(proto): add physical hashJoin & mergeJoin physical relations

### PR DESCRIPTION
**Pull request by chaojun-zhang**
This PR includes hashJoin and mergeJoin physical plan relations according the spec

### CODE REVIEW

**Comment by chaojun-zhang** [comment]
@jvanstraten , sorry for my mistake  that  close the original PR , I can't reopen it so raise a new PR here

**Comment by westonpace** [comment]
Thank you for contributing these.  I'm curious about the duplication here, though I don't know if it's a bad thing entirely.

In #331 we decided that trying too hard to avoid duplication (e.g. by basing as-of join on a logical join) was probably a bad idea.  However, in this PR we are providing direct physical specializations of the logical join.  Do you think it would make sense to use inheritance here?  Is there a reason you chose not to?  Can you imagine the physical specializations diverging from the logical specifications (other than just adding on new fields) at some point?

**Comment by jvanstraten** [comment]
I don't know if this is true, but if all physical joins are likely to use the same fields, shouldn't the physical implementation just be an optional field on `JoinRel`?

If not, for some of the cases like this for the validator-specific protos (especially scalar vs aggregate vs window functions) I made separate messages for the common fields, in this case for example `JoinCommon` or something. That way, if we add logical-specific or physical-specific fields later, we can do that outside that common structure, but also avoid duplication. It would be a breaking change for `JoinRel` to add this indirection, though.

**Comment by chaojun-zhang** [comment]
To avoid duplicate, I think we may have two options: 

- option 1: add new field inside JoinRel like below to distinguish logical join , hash join , merge join  and nested loop join
 ```
JoinKind kind = 7;

  enum JoinKind {
    JOIN_KIND_LOGICAL = 0;
    JOIN_KIND_HASH = 1;
    JOIN_KIND_MERGE = 2;
    JOIN_KIND_NESTED_LOOP = 3;
  }
```
- option 2: extract  join fields to JoinCommon message, and use inheritance for logical join ,hash join ,merge join and nested loop join 

I'm not sure which option is prefer to,  my thinking the option 1  may be better  since these join operator share the same attributes  according to the spec in substrait.io?

File: proto/substrait/algebra.proto
Comment (jacques-n) [inline proto/substrait/algebra.proto:490]: I think we should specialize these fields for what a dumb consumer would want. I think that is probably:

FieldReference[] left_keys (hash table)
FieldReference[] left_keys
Expression during_join_filter (join conditions)
Expression post_join_filter (afterwards)

We should also validate that the join types are valid in this situation.

File: proto/substrait/algebra.proto
Comment (jacques-n) [inline proto/substrait/algebra.proto:517]: Same as above, I think we need to specialize for dumb consumers when producing physical rels like this.

**Review by jacques-n — CHANGES_REQUESTED** [review]
In general, I think we should avoid collapsing these. I think they actually need to be different. They have different physical property maintenance characterististics and will probably develop in different directions.

Once we get a good definition, let's make sure we also update the associated site docs.

File: proto/substrait/algebra.proto
Comment (westonpace) [reply proto/substrait/algebra.proto]: I like this idea.  @jvanstraten @vibhatha and I have struggled at times with the mismatch that Acero expects simple field references (except in the project relation) in many places the official spec contains full expressions.  This comment helped me crystallize that this is because those spots with full expressions are simply more logical than physical.

Comment (jvanstraten) [reply proto/substrait/algebra.proto]: +1; I don't personally know what it should look like, but for physical relations I wasn't expecting expressions at all. I'd go as far as to say that the and-tree-of-equalities expression should be entirely reduced to just a bunch of FieldReferences, so a consumer doesn't have to match and unpack something involving something as complex as extension functions just to understand what the producer meant.

Comment (westonpace) [reply proto/substrait/algebra.proto]: > the and-tree-of-equalities expression should be entirely reduced to just a bunch of FieldReferences

For reference, in Acero, we currently only support a vector of field references (one from the left and one from the right) and, for each field, an enum of `equal` or `is_not_distinct_from`.  Though that would limit this to equi-joins.

File: proto/substrait/algebra.proto
Comment (chaojun-zhang) [reply proto/substrait/algebra.proto]: updated

File: proto/substrait/algebra.proto
Comment (chaojun-zhang) [reply proto/substrait/algebra.proto]: Reading  [velox ](https://github.com/facebookincubator/velox)code,  hash join and merge join also support a vector of field references for 'lefyKeys' and 'rightKeys', but how can I pass the [outputType ](https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.h#L1086)from proto , how can I pass it from proto ?

File: proto/substrait/algebra.proto
Comment (chaojun-zhang) [reply proto/substrait/algebra.proto]: updated

File: proto/substrait/algebra.proto
Comment (jacques-n) [reply proto/substrait/algebra.proto]: I think it is valuable to support the equal/is not distinct concepts for each of the key pairings here. Probably makes most sense to use a message similar to what it sounds like acero does (e.g. left_key, right_key, equality_condition)

WRT to join type, I'm inclined to only introduce the join types people are using for each of these. We can introduce more as people have them. As such, I'd suggest we have separate enums for each (HashJoinType, MergeJoinType, etc). What do others think?

Comment (jvanstraten) [reply proto/substrait/algebra.proto]: > As such, I'd suggest we have separate enums for each (HashJoinType, MergeJoinType, etc). What do others think?

I agree with the thought because they seem different enough (and also agree with your other paragraph). However, I'd personally opt against adding new enum types in the protos, because enums in general are pretty fragile compared to everything else in protobuf (lack of proper scoping, annoyingly long names in JSON, representing unknown values with invalid enum values in C, easy for consumers to forget to handle unrecognized values, can't add options that contain values later like you can with e.g. Rust enums, etc). A message containing a `oneof`s with (initially) empty messages seems superior to me in every way. I don't think it even costs more in the binary serialization.

Comment (jacques-n) [reply proto/substrait/algebra.proto]: I'm actually pretty supportive of this pattern generally. It feels much more like rust enums.

File: site/docs/relations/physical_relations.md
Comment (westonpace) [inline site/docs/relations/physical_relations.md:23]: Since this is a physical relation do we want to explicitly declare that the left side is the build side and the right side is the probe side (or vice versa)?  It looks like you are already making this assumption in your description of left keys and right keys.

File: site/docs/relations/physical_relations.md
Comment (westonpace) [inline site/docs/relations/physical_relations.md:24]:
```suggestion
 | Left Keys           | References to the fields to join on in the left input.                                                                                                                                                                                                                                                 | Optional, defaults to blank if during join predicate is specified          |
```

Same change for right keys

File: site/docs/relations/physical_relations.md
Comment (westonpace) [inline site/docs/relations/physical_relations.md:26]: I don't see this in the proto definition

File: proto/substrait/algebra.proto
Comment (westonpace) [reply proto/substrait/algebra.proto]: > updated, do we need to extract JoinType as shared message?


I think that would be a good idea.

**Review by westonpace — COMMENTED** [review]
This review is long overdue.  I have a few more questions.

File: site/docs/relations/physical_relations.md
Comment (chaojun-zhang) [reply site/docs/relations/physical_relations.md:23]: I think we can leave it to the user to decide which side is probe or build , your thoughts?

File: site/docs/relations/physical_relations.md
Comment (chaojun-zhang) [reply site/docs/relations/physical_relations.md:26]: Removed

File: site/docs/relations/physical_relations.md
Comment (westonpace) [reply site/docs/relations/physical_relations.md]: Picking the correct side is generally an optimizer task (requires statistics / heuristics).  My assumption for more physical nodes is that they've already gone through an optimizer.

For a single optimizer/consumer pair this wouldn't be a problem.  They can both agree offline which side should be build and probe.  However, imagine an engine-independent substrait optimizer created this plan and hoped for it to be consumed by many different engines.  In that case the different consumers might pick left/right build/probe differently.

That being said, I haven't worked much with optimizers.  CC @jacques-n who has more experience here.

**Comment by chaojun-zhang** [comment]
Please ignore this PR since lots of commit message can't pass the Lint check

**Comment by chaojun-zhang** [comment]
@jacques-n @westonpace Please take a further review and help trigger the check again.

File: site/docs/relations/physical_relations.md
Comment (westonpace) [inline site/docs/relations/physical_relations.md:23]: Either we should assign these as probe and build (my preference would be to be explicit) or we should specifically explain that consumers should assign probe & build based on whatever criteria they want.  As it is, I think this is ambiguous.

Comment (chaojun-zhang) [reply site/docs/relations/physical_relations.md]: I found that the documentation has mentioned that the right input is build-side, see [this](https://substrait.io/relations/physical_relations/#hash-equijoin-operator)

I also updated in this section [https://substrait.io/relations/physical_relations/#hash-equijoin-properties]

**Comment by chaojun-zhang** [comment]
@westonpace I have update the document and help trigger the check again.

**Comment by westonpace** [comment]
:( It seems our release script is failing (unrelated to this change) I will see if I can fix that real quick.

**Comment by westonpace** [comment]
Fixed.  Please rebase and I will run again.

**Review by westonpace — DISMISSED** [review]

**Comment by westonpace** [comment]
This looks good to me.  Will also need approval from @jacques-n or Jacques

**Comment by chaojun-zhang** [comment]
@westonpace @jacques-n I have rebased and please help trigger the check again

**Comment by chaojun-zhang** [comment]
@westonpace PR rebased, Please take a look again. Thanks!

**Review by jacques-n — APPROVED** [review]
LGTM. With @westonpace approval, I'm going to merge.

**Pull request merged**
